### PR TITLE
Fixing the amount by correct handling the number of BigUint decimals

### PR DESCRIPTION
**Pull request by alyn509**

### CODE REVIEW

File: bridged-tokens-wrapper/src/lib.rs [hunk 1]
```diff
@@ -132,23 +186,27 @@ pub trait BridgedTokensWrapper {
             payment_token == universal_bridged_token_ids,
             "Esdt token unavailable"
         );
-
+        self.require_tokens_have_set_decimals_num(&payment_token, &requested_token);
         let chain_specific_token_id = &requested_token;
+        let universal_token_decimals = self.token_decimals_num(&payment_token).get();
```
Comment (sasurobert) [inline bridged-tokens-wrapper/src/lib.rs:191]: you could make an internal function for the conversion and call the same function on both wrap and unwrap functions.

File: bridged-tokens-wrapper/src/dfp_big_uint.rs [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+            let diff_decimals = self.num_decimals - decimals;
+            return DFPBigUint {
+                bu: self.bu.clone() / 10u32.pow(diff_decimals),
+                num_decimals: decimals,
+            };
+        }
+        self.clone()
+    }
+
+    pub fn trunc(&self) -> Self {
```
[the remark this answers is not in the document]
Comment (alyn509) [reply bridged-tokens-wrapper/src/dfp_big_uint.rs:33]: for test purposes was introduced. The whole struct will be moved in the framework so don't worry about it

**Review by sasurobert — COMMENTED** [review]
still some tests are failing

**Comment by alyn509** [comment]
> still some tests are failing

**Review by sasurobert — APPROVED** [review]

File: .vscode/launch.json [hunk 3]
```diff
@@ -1,12 +1,58 @@
 {
+    // Use IntelliSense to learn about possible attributes.
```
Comment (sasurobert) [inline .vscode/launch.json:2]: delete commented stuff. Why do we need .vscode folder here ?

**Review by sasurobert — APPROVED** [review]

**Review by dragos-rebegea — APPROVED** [review]

File: bridged-tokens-wrapper/mandos/blacklist_token.scen.json [hunk 4]
```diff
@@ -109,8 +109,10 @@
                         "str:universalBridgedTokenIds.len": "1",
                         "str:universalBridgedTokenIds.index|nested:str:WUSDC-abcdef": "1",
                         "str:universalBridgedTokenIds.item|u32:1": "str:WUSDC-abcdef",
-                        "str:tokenLiquidity|nested:str:USDC-aaaaaa": "200",
-                        "str:tokenLiquidity|nested:str:USDC-cccccc": "100"
+                        "str:tokenLiquidity|nested:str:USDC-aaaaaa": "200000000000000",
+                        "str:tokenLiquidity|nested:str:USDC-cccccc": "100000000000000",
+                        "str:token_decimals_num|nested:str:WUSDC-abcdef": "0x06",
```
Comment (andrei-marinica) [inline bridged-tokens-wrapper/mandos/blacklist_token.scen.json:114]: Values can be expressed in base 10, i.e. 6 and 18, respectively.

File: bridged-tokens-wrapper/src/dfp_big_uint.rs [hunk 5]
```diff
@@ -0,0 +1,43 @@
+elrond_wasm::imports!();
+elrond_wasm::derive_imports!();
+
+#[derive(Clone, PartialEq, Eq)]
+pub struct DFPBigUint<M: ManagedTypeApi> {
```
Comment (andrei-marinica) [inline bridged-tokens-wrapper/src/dfp_big_uint.rs:5]: Add a comment saying that this is a Decimal Fixed Point Big Uint.

**Review by andrei-marinica — DISMISSED** [review]

File: bridged-tokens-wrapper/src/dfp_big_uint.rs [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+        }
+        if self.num_decimals > decimals {
+            let diff_decimals = self.num_decimals - decimals;
+            return DFPBigUint {
+                bu: self.bu.clone() / 10u64.pow(diff_decimals),
+                num_decimals: decimals,
+            };
+        }
+        self.clone()
+    }
```
Comment (iancudorinmarian) [inline bridged-tokens-wrapper/src/dfp_big_uint.rs:31]: This function can be made more efficient:
- consume `self` and let the caller clone if they need to
- use early exit conditions (it's the most common, and it also makes the code more readable)

```suggestion
    pub fn convert(self, decimals: u32) -> Self {
            if self.num_decimals == decimals {
                 return self;
            }
    
        let new_bu=  if self.num_decimals > decimals { 
             let diff_decimals = self.num_decimals - decimals;
             self.bu / 10u64.pow(diff_decimals)
         else { 
               let diff_decimals = decimals - self.num_decimals;
               self.bu * 10u64.pow(diff_decimals)
         };
         
         DFPBigUint {
                bu: new_bu,
                num_decimals: decimals,
            }
```

File: bridged-tokens-wrapper/src/lib.rs [hunk 7]
```diff
@@ -99,13 +150,21 @@ pub trait BridgedTokensWrapper {
             // if there is chain specific -> universal mapping, then the token is whitelisted
             let new_payment = if !universal_token_id_mapper.is_empty() {
```
Comment (iancudorinmarian) [inline bridged-tokens-wrapper/src/lib.rs:151]: de-nest some of my old code:
```
if universal_token_id_mapper.is_empty() {  
       new_payments.push(payment); 
      continue; 
}

// the code in the original `if` here

new_payments.push(new_payment);
```

File: bridged-tokens-wrapper/src/lib.rs [hunk 8]
```diff
@@ -132,23 +191,42 @@ pub trait BridgedTokensWrapper {
...
+
+    fn get_converted_amount(
+        &self,
+        from: &TokenIdentifier,
+        to: &TokenIdentifier,
+        amount: BigUint,
+    ) -> BigUint {
+        let from_decimals = self.token_decimals_num(from).get();
+        let to_decimals = self.token_decimals_num(to).get();
+        let amount = DFPBigUint::from_raw(amount, from_decimals);
```
Comment (iancudorinmarian) [inline bridged-tokens-wrapper/src/lib.rs:228]: please don't shadow the original `amount` argument. Rename this.

**Review by sasurobert — DISMISSED** [review]

File: bridged-tokens-wrapper/src/lib.rs [hunk 9]
```diff
@@ -97,20 +148,31 @@ pub trait BridgedTokensWrapper {
                 self.chain_specific_to_universal_mapping(&payment.token_identifier);
 
             // if there is chain specific -> universal mapping, then the token is whitelisted
-            let new_payment = if !universal_token_id_mapper.is_empty() {
+            if universal_token_id_mapper.is_empty() {
+                new_payments.push(payment);
+                continue;
+            } else {
```
Comment (iancudorinmarian) [inline bridged-tokens-wrapper/src/lib.rs:154]: Remove the `else` entirely, you have `continue` in the `if`

**Review by iancudorinmarian — DISMISSED** [review]

**Review by andreibancioiu — CHANGES_REQUESTED** [review]
Cargo lock is missing.

**Review by sasurobert — APPROVED** [review]

**Review by andreibancioiu — COMMENTED** [review]
Cargo lock present on the contract of interest :+1:

**Review by andreibancioiu — APPROVED** [review]

**Pull request merged**